src/wasm.h

```diff
@@ -1137,61 +1137,62 @@ class AllocatingModule : public Module {
-  virtual ReturnType visitBreak(Break *curr) { abort(); }
-  virtual ReturnType visitSwitch(Switch *curr) { abort(); }
-  virtual ReturnType visitCall(Call *curr) { abort(); }
-  virtual ReturnType visitCallImport(CallImport *curr) { abort(); }
-  virtual ReturnType visitCallIndirect(CallIndirect *curr) { abort(); }
-  virtual ReturnType visitGetLocal(GetLocal *curr) { abort(); }
-  virtual ReturnType visitSetLocal(SetLocal *curr) { abort(); }
-  virtual ReturnType visitLoad(Load *curr) { abort(); }
-  virtual ReturnType visitStore(Store *curr) { abort(); }
-  virtual ReturnType visitConst(Const *curr) { abort(); }
-  virtual ReturnType visitUnary(Unary *curr) { abort(); }
-  virtual ReturnType visitBinary(Binary *curr) { abort(); }
-  virtual ReturnType visitSelect(Select *curr) { abort(); }
-  virtual ReturnType visitHost(Host *curr) { abort(); }
-  virtual ReturnType visitNop(Nop *curr) { abort(); }
-  virtual ReturnType visitUnreachable(Unreachable *curr) { abort(); }
+  ReturnType visitBlock(Block *curr) { abort(); }
+  ReturnType visitIf(If *curr) { abort(); }
+  ReturnType visitLoop(Loop *curr) { abort(); }
+  ReturnType visitBreak(Break *curr) { abort(); }
+  ReturnType visitSwitch(Switch *curr) { abort(); }
+  ReturnType visitCall(Call *curr) { abort(); }
+  ReturnType visitCallImport(CallImport *curr) { abort(); }
+  ReturnType visitCallIndirect(CallIndirect *curr) { abort(); }
+  ReturnType visitGetLocal(GetLocal *curr) { abort(); }
+  ReturnType visitSetLocal(SetLocal *curr) { abort(); }
+  ReturnType visitLoad(Load *curr) { abort(); }
+  ReturnType visitStore(Store *curr) { abort(); }
+  ReturnType visitConst(Const *curr) { abort(); }
+  ReturnType visitUnary(Unary *curr) { abort(); }
+  ReturnType visitBinary(Binary *curr) { abort(); }
+  ReturnType visitSelect(Select *curr) { abort(); }
+  ReturnType visitHost(Host *curr) { abort(); }
+  ReturnType visitNop(Nop *curr) { abort(); }
+  ReturnType visitUnreachable(Unreachable *curr) { abort(); }
   // Module-level visitors
-  virtual ReturnType visitFunctionType(FunctionType *curr) { abort(); }
-  virtual ReturnType visitImport(Import *curr) { abort(); }
-  virtual ReturnType visitExport(Export *curr) { abort(); }
-  virtual ReturnType visitFunction(Function *curr) { abort(); }
-  virtual ReturnType visitTable(Table *curr) { abort(); }
-  virtual ReturnType visitMemory(Memory *curr) { abort(); }
-  virtual ReturnType visitModule(Module *curr) { abort(); }
+  ReturnType visitFunctionType(FunctionType *curr) { abort(); }
+  ReturnType visitImport(Import *curr) { abort(); }
+  ReturnType visitExport(Export *curr) { abort(); }
+  ReturnType visitFunction(Function *curr) { abort(); }
+  ReturnType visitTable(Table *curr) { abort(); }
+  ReturnType visitMemory(Memory *curr) { abort(); }
+  ReturnType visitModule(Module *curr) { abort(); }
 
   ReturnType visit(Expression *curr) {
     assert(curr);
+    SubType* self = static_cast<SubType*>(this);
```

Comment:
Yes, that's the ugly bit but it's squirreled away in the implementation. LLVM does it with a macro:
https://github.com/llvm-mirror/llvm/blob/868145efb053c3f9294676cf4f36d6220e500269/include/llvm/IR/InstVisitor.h#L30

I think it's pretty nice with a macro :-)